directives/run-async/run-async.js

```diff
@@ -11,16 +16,103 @@
  * http://polymer.github.io/PATENTS.txt
  */
+import { AsyncDirective } from 'lit-html/async-directive.js';
 import { AsyncStateEvent } from '../../helpers/asyncStateEvent.js';
-import { directive } from 'lit-html/lit-html.js';
+import { directive } from 'lit-html/directive.js';
+import { noChange } from 'lit-html';
 
 const hasAbortController = typeof AbortController === 'function';
-const runs = new WeakMap();
+
 /**
  * Error thrown by async tasks when the task couldn't be started based on the
  * key passed to it.
  */
 export class InitialStateError extends Error {}
+
+class RunAsync extends AsyncDirective {
+
+	constructor(partInfo) {
+		super(partInfo);
+		this.abortController = hasAbortController ? new AbortController() : undefined;
+		this.key = undefined;
+		this.pendingPromise = undefined;
+		this.rejectPending = undefined;
+		this.resolvePending = undefined;
+		this.state = 'initial';
+	}
+
+	render(key, task, templates) {
+
+		const { success, pending, initial, failure } = templates;
+
+		// first time we see a value we save and await the work function.
+		if (this.key !== key) {
+
+			// abort a pending request if there is one
+			if (this.state === 'pending') {
+				if (this.abortController !== undefined) {
+					this.abortController.abort();
+				}
+			}
+
+			const taskPromise = task(key, {
+				signal: hasAbortController ? this.abortController.signal : undefined
+			});
+			// The state is immediately 'pending', since the function has been
+			// executed, but if the function throws an InitialStateError to
+			// indicate that it couldn't even start processing, then we will set
+			// the state to 'initial'.
+			this.key = key;
+			this.pendingPromise = new Promise((res, rej) => {
+				this.resolvePending = res;
+				this.rejectPending = rej;
+			}).catch((error) => {
+				// swallow initial state errors
+				if (!(error instanceof InitialStateError)) {
+					throw error;
+				}
+			});
+			this.state = 'pending';
+			Promise.resolve(taskPromise).then((value) => {
+				this.state = 'success';
+				this.resolvePending();
+				this.setValue(success(value));
+			}, (error) => {
+				this.rejectPending(error);
+				if (error instanceof InitialStateError && typeof initial === 'function') {
+					this.state = 'initial';
+					this.setValue(initial());
+				} else {
+					this.state = 'failure';
+					if (typeof failure === 'function') {
+						this.setValue(failure(error));
+					}
+				}
+			});
+		}
+
+		if ((this.state === 'pending') && typeof pending === 'function') {
+			return pending();
+		}
+
+		return noChange;
+
+	}
+
+	update(part, [key, task, templates, options]) {
+		const renderResult = this.render(key, task, templates);
+		const directiveOptions = Object.assign({ pendingState: true }, options);
+		if (directiveOptions.pendingState && this.key === key && this.state === 'pending') {
+			const element = part.parentNode.nodeType === Node.ELEMENT_NODE ? part.parentNode : part.parentNode.host;
+			if (element) {
```

Comment:
Could `element` ever be `null` or `undefined` and `update` not run again, such that the event never gets dispatched?

Reply:
Or, what if the async task immediately returns a resolve promise and the state goes from pending -> success?  Could that happen before this `update` method runs, where dispatching the event relies on the state change?  I'm not sure what the behaviour of that was before.

Reply:
I think that's what's happening in Firefox & legacy-Edge in the failing tests. :(